### PR TITLE
.circleci/config.yml: Update CircleCI syntax

### PR DESCRIPTION
**Pull request by harshhx17**
This updates CircleCI syntax from 1.0 to 2.0 and fixes uncached
build failures due to deps.opam.sh

Closes https://github.com/coala/coala-bears/issues/2298
Closes https://github.com/coala/coala-bears/issues/2006

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by harshhx17** [comment]
I needed some help 
Using multiple languages in a build is not supported directly in CircleCi 2.0
A probable solution will be to use a custom docker image
There is also a way in which we can use different dockers for different parts of the build ....
But I don't think that will be useful as builds from one docker will not stay in the other docker
And the tests or commands that require more than one language will still be a problem
and we will continuously need to change the CircleCi config with new tests..
If we decide on writing a new docker image too....
Where should we host it??

File: .ci/deps.pip.sh [hunk 1]
```diff
@@ -13,7 +13,7 @@ esac
 
 for dep_version in "${dep_versions[@]}" ; do
   pyenv install -ks $dep_version
-  pyenv local $dep_version 2.7.10
+  pyenv local $dep_version #2.7.10
```
Comment (jayvdb) [inline .ci/deps.pip.sh]: err, it cant get merged like that.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            - "~/.ghc"
+            - "~/.ghc-mod"
+            - "~/infer-linux64-v0.7.0"
+            - "~/nltk_data"
+            - "~/pmd-bin-5.4.1"
+            - "~/.local/share/coala-bears"
+            - "~/bakalint-0.4.0"
+            - "~/.julia"
+            - "~/.local/tailor/"
+            - "~/.venv"
```
Comment (jayvdb) [inline .circleci/config.yml]: missing EOL at EOF

File: .ci/deps.pip.sh (hunk 1 above)
Comment (harshhx17) [reply .ci/deps.pip.sh]: Sorry, I was just testing and on using python image for circleci2.0,
Python version 2.7.10 was not installed itself...
So, I just commented the version

File: .circleci/config.yml (hunk 2 above)
Comment (harshhx17) [reply .circleci/config.yml]: will add that :wink:

**Comment by harshhx17** [comment]
Sure @jayvdb :smile:

**Comment by jayvdb** [comment]
Circle failed.

**Comment by harshhx17** [comment]
@jayvdb 
This is the same error that I was getting on https://github.com/coala/coala-bears/pull/2292 when it was running on my fork....
Should I open an issue for it?

**Comment by harshhx17** [comment]
All the dependencies are finally installed but the tests are failing and need to remove the code redundancy.
Just made a pr to see if I am going in the right direction :smile:

**Comment by jayvdb** [comment]
@harshhx17 you probably want to help with https://github.com/coala/coala/pull/5234 ; that is a much simpler case, and once that is merged we'll start learning more about the new Circle, which will help with this PR.

**Comment by harshhx17** [comment]
Sorry for replying this late and being inactive for these 5 days...
Actually, I was in Dubai where I did not have a laptop or internet connectivity..
Can you please help me figure out how I should help with https://github.com/coala/coala/pull/5234 ... I think it is already assigned to alphadose and the builds are passing

**Comment by harshhx17** [comment]
Meanwhile, should I stop working on this issue ?

**Comment by jayvdb** [comment]
coala is now using circle 2.0
@harshhx17 , your turn

**Comment by harshhx17** [comment]
Sure...I am pushing the same changes as the last time for once as its builds were passing on my fork.
So that we can discuss why the tests are failing... :sunglasses:

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    parallelism: 2
+
+    steps:
+      - checkout
+      - restore_cache:
+          key: dependency-cache-{{checksum "requirements.txt"}}
+      - run:
+          name: dependencies
+          command: |
+                    sudo apt-get update
```
Comment (sangamcse) [inline .circleci/config.yml]: Why this much indentation?

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    rvm use 2.2.2 --default
+                    ruby -v
+                    gem install bundler
+                    bash .ci/deps.sh
+      - run:
+         name: CabalDependencies
+         command:  bash .ci/deps.cabal.sh
+      - run:
+         name: DependenciesPip
+         command:   |
```
Comment (sangamcse) [inline .circleci/config.yml]: Extra space here

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    docker:
+      - image: circleci/jdk8:0.1.1
+
+    parallelism: 2
+
+    steps:
+      - checkout
+      - restore_cache:
+          key: dependency-cache-{{checksum "requirements.txt"}}
+      - run:
```
Comment (sangamcse) [inline .circleci/config.yml]: Keep one blank line here

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                    sudo apt-get update
+                    sudo apt-get install -y python3.4-venv git bzip2 \
+                      cabal-install-1.24 clang-3.4 libreadline-dev libyaml-dev\
+                      gfortran ghc happy indent libblas-dev liblapack-dev \
+                      libperl-critic-perl libxml2-utils menhir php-codesniffer \
+                      build-essential ruby texinfo libbz2-dev libcurl4-openssl-dev \
+                      libexpat-dev libncurses-dev zlib1g-dev git-core libssl-dev \
+                      libsqlite3-dev sqlite3 libxslt1-dev libffi-dev yarn python3-dev \
+                      libgdbm-dev libncurses5-dev automake libtool bison
+      - run:
```
Comment (sangamcse) [inline .circleci/config.yml]: One blank here too

Comment (sangamcse) [reply .circleci/config.yml]: Correct all indentation and blank line before `- run`

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                      export NVM_DIR="$HOME/.nvm"
+                      [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+                      pip install $(ls ./dist/*.whl)"[alldeps]"
+      - run:
+          name: Documents
+          when: always
+          command: |
+                      echo $CIRCLE_NODE_INDEX
+                      if [ $CIRCLE_NODE_INDEX == 1 ] ; then
+                      pip install -r docs-requirements.txt
```
Comment (sangamcse) [inline .circleci/config.yml]: 2 space indentation here

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,170 @@
+version: 2
+jobs:
+  build:
+    working_directory: ~/coala-bears
+
+    docker:
+      - image: circleci/jdk8:0.1.1
+
+    parallelism: 2
```
Comment (alphadose) [inline .circleci/config.yml]: This line is redundant. Instead of parallelism use workflows https://circleci.com/docs/2.0/workflows/.

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                    echo 'export PATH=$HOME/dart-sdk/bin:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/.cabal/bin:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/infer-linux64-v0.7.0/infer/bin:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/pmd-bin-5.4.1/bin:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/bakalint-0.4.0:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/elm-format-0.18:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/.local/tailor/tailor-latest/bin:$PATH' >> $BASH_ENV
+                    echo 'export PATH=$HOME/phpmd:$PATH' >> $BASH_ENV
+                    echo 'export R_LIB_USER=$HOME/.RLibrary:$PATH' >> $BASH_ENV
+      - run:
```
Comment (alphadose) [inline .circleci/config.yml]: Always give one blank line between the different blocks under **steps** . Increases the readability.

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+          name: GeneralDependencies
+          command: |
+                    gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
+                    curl -sSL https://get.rvm.io | bash -s stable
+                    source /etc/profile.d/rvm.sh
+                    rvm install 2.2.2
+                    rvm use 2.2.2 --default
+                    ruby -v
+                    gem install bundler
+                    bash .ci/deps.sh
```
Comment (alphadose) [inline .circleci/config.yml]: The proper way to indent is 2 spaces under each block. Fix that in each of the sub blocks.

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+         name: CabalDependencies
+         command:  bash .ci/deps.cabal.sh
+      - run:
+         name: DependenciesPip
+         command:   |
+                    git clone https://github.com/pyenv/pyenv.git /tmp/pyenv
+                    if [ ! -d "$HOME/.pyenv" ]; then mkdir -p ~/.pyenv ; fi
+                    cp -R /tmp/pyenv/* ~/.pyenv
+                    export PYENV_ROOT="$HOME/.pyenv"
+                    export PATH="$PYENV_ROOT/bin:$PATH"
```
Comment (alphadose) [inline .circleci/config.yml]: Setup all the bash environment variables under 1 block only.

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+         command: bash .ci/deps.java.sh
+      - run:
+         name: OpamDependencies
+         command:   bash .ci/deps.opam.sh
+      - run:
+         name: RDependencies
+         command:   bash .ci/deps.r.sh
+      - run:
+         name: CoalaDependencies
+         command:   bash .ci/deps.coala-bears.sh
```
Comment (alphadose) [inline .circleci/config.yml]: Give only 1 space indentation for all of the above single line commands.

File: .circleci/config.yml [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+          name: Documents
+          when: always
+          command: |
+                      echo $CIRCLE_NODE_INDEX
+                      if [ $CIRCLE_NODE_INDEX == 1 ] ; then
+                      pip install -r docs-requirements.txt
+                      fi
+      - run:
+          name: SetupDocs
+          when: always
```
Comment (alphadose) [inline .circleci/config.yml]: Remove this line ( when: always). This command shouldn't run if the dependency and environment setup fails.

File: .circleci/config.yml [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+          command: |
+                      echo $CIRCLE_NODE_INDEX
+                      if [ $CIRCLE_NODE_INDEX == 1 ] ; then
+                        python setup.py docs
+                      fi
```
Comment (alphadose) [inline .circleci/config.yml]: There is no need to use hacks using $CIRCLE_NODE_INDEX in 2.0. Workflows solved this problem. See my above comment.

**Comment by alphadose** [comment]
Had a quick look at your log. **pyenv not found** is one of the problems. You need to export the directory where pyenv executable is located to $PATH. 
export $PATH=$PATH:/path/to/pyenv

File: .circleci/config.yml (hunk 6 above)
Comment (harshhx17) [reply .circleci/config.yml]: Sure sangam, I will correct all the indentation in the next push
Thanks

File: .circleci/config.yml (hunk 8 above)
Comment (harshhx17) [reply .circleci/config.yml]: Thanks alphadose, will definately  look into it

File: .circleci/config.yml (hunk 14 above)
Comment (harshhx17) [reply .circleci/config.yml]: Oh! That had be better...

**Comment by harshhx17** [comment]
Will try that in my fork

**Comment by alphadose** [comment]
@harshhx17 I believe the tests are failing or being skipped due to dependency issues. coala-bears has a lot of dependencies and I wouldn't be surprised if some of them were not installed properly. You can see the coverage reports and see which lines are not covered. I can't go into details but my guess is that within those lines those dependencies were called or required. If you go through those lines, maybe you can trace back and find the root of this problem.

**Comment by jayvdb** [comment]
You are doing too many builds.

And you need to solve https://github.com/coala/coala-bears/issues/2006 first

File: .ci/deps.opam.sh [hunk 15]
```diff
@@ -1,13 +1,14 @@
 set -e
 set -x
 
-TERM=dumb
+export TERM=dumb
```
Comment (jayvdb) [inline .ci/deps.opam.sh]: unrelated change.

what is this fixing?

create an issue about it.

File: .ci/deps.opam.sh [hunk 16]
```diff
@@ -1,13 +1,14 @@
...
 
 # Infer commands
 if [ ! -e ~/infer-linux64-v0.7.0/infer/bin ]; then
   wget -nc -O ~/infer.tar.xz https://github.com/facebook/infer/releases/download/v0.7.0/infer-linux64-v0.7.0.tar.xz
   tar xf ~/infer.tar.xz -C ~/
   cd ~/infer-linux64-v0.7.0
+  apt-get install ocaml ocaml-native-compilers camlp4-extra opam -y
```
Comment (jayvdb) [inline .ci/deps.opam.sh]: is this the fix for https://github.com/coala/coala-bears/issues/2006 ?

File: .ci/deps.opam.sh (hunk 15 above)
Comment (harshhx17) [reply .ci/deps.opam.sh]: I had faced an error saying TERM not found...
This change solved it...
Will try to find the link to that log

File: .ci/deps.opam.sh (hunk 16 above)
Comment (harshhx17) [reply .ci/deps.opam.sh]: This is not the complete fix,
There were also some other problems relating to the ubuntu-12.04 image in the current circle where we had used:
if [ -n "$ADD_APT_UBUNTU_RELEASE" ]; then
  echo "deb http://archive.ubuntu.com/ubuntu/ $ADD_APT_UBUNTU_RELEASE main universe" | sudo tee -a /etc/apt/sources.list.d/$ADD_APT_UBUNTU_RELEASE.list > /dev/null

File: .ci/deps.opam.sh (hunk 15 above)
Comment (jayvdb) [reply .ci/deps.opam.sh]: ping?  This would be a separate bug...

File: .ci/deps.pip.sh [hunk 17]
```diff
@@ -4,17 +4,17 @@ set -x
...
- 1) dep_versions=( "3.4.3" ) ;;
- -1) dep_versions=( ) ;;  # set by .travis.yml
- *) dep_versions=( "3.5.1" ) ;;
+case $CIRCLE_JOB in
+  "python-3.5") dep_versions=( "3.4.3" "3.5.1" );;
+  "python-3.4") dep_versions=( "3.4.3" );;
+  *)  dep_versions=( "3.5.1" );;
 esac
 
+pyenv install -ks 2.7.10
```
Comment (jayvdb) [inline .ci/deps.pip.sh]: why installing 2.7.10?

File: .ci/deps.pip.sh [hunk 18]
```diff
@@ -4,17 +4,17 @@ set -x
...
+
 for dep_version in "${dep_versions[@]}" ; do
   pyenv install -ks $dep_version
   pyenv local $dep_version 2.7.10
-  python --version
```
Comment (jayvdb) [inline .ci/deps.pip.sh]: undo this!

this is how we know the correct version will be used.

File: .ci/deps.sh [hunk 19]
```diff
@@ -58,7 +58,7 @@ if [ ! -e ~/phpmd/phpmd ]; then
 fi
 
 # astyle installation
-wget "https://downloads.sourceforge.net/project/astyle/astyle/astyle%203.0.1/astyle_3.0.1_linux.tar.gz?r=&ts=1499017588&use_mirror=excellmedia" -O ~/astyle.tar.gz
+wget "https://downloads.sourceforge.net/project/astyle/astyle/astyle%203.0.1/astyle_3.0.1_linux.tar.gz?r=&ts=1499017588&use_mirror=excellmedia" -O ~/astyle.tar.gz --no-check-certificate
```
Comment (jayvdb) [inline .ci/deps.sh]: why this?
surely that would be a separate commit.

File: .circleci/config.yml [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+            - run:
+                name: dependencies
+                command: |
+                  sudo apt-get update
+                  sudo apt-get install -y software-properties-common python-software-properties curl
+                  sudo add-apt-repository -y ppa:hvr/ghc
+                  sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+                  sudo add-apt-repository -y ppa:avsm/ppa
+                  curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash  -
+                  curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
```
Comment (jayvdb) [inline .circleci/config.yml]: why using yarn?

We have another issue about yarn.

I can re-assign that to you, if you explain why it was necessary for Circle 2.0

File: .circleci/config.yml [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+                  libexpat-dev libncurses-dev zlib1g-dev git-core libssl-dev \
+                  libsqlite3-dev sqlite3 libxslt1-dev libffi-dev yarn python3-dev \
+                  libgdbm-dev libncurses5-dev automake libtool bison \
+                  libncursesw5-dev libc6-dev tk-dev
+
+            - run:
+                name: virtual-env
+                command: |
+                  if [ ! -d "$HOME/.venv" ]; then
+                  python3 -m venv ~/.venv
```
Comment (jayvdb) [inline .circleci/config.yml]: why do you need a venv here, if you are using `pyenv` below?

File: .circleci/config.yml [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+                  gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
+                  curl -sSL https://get.rvm.io | bash -s stable
+                  source /etc/profile.d/rvm.sh
+                  echo 'source /etc/profile.d/rvm.sh' >> $BASH_ENV
+                  rvm install 2.2.2
+                  rvm use 2.2.2 --default
+                  ruby -v
+                  gem install bundler
+                  bash .ci/deps.sh
+                  bundle install
```
Comment (jayvdb) [inline .circleci/config.yml]: `.ci/deps.sh` already does a `bundle install`.  Why do you need another one here?

File: .circleci/config.yml [hunk 23]
```diff
@@ -0,0 +1,188 @@
+version: 2
+jobs:
+    python-3.5: &build
+        docker:
+            - image: circleci/jdk8:0.1.1
```
Comment (jayvdb) [inline .circleci/config.yml]: why this image?

File: .ci/deps.pip.sh (hunk 18 above)
Comment (harshhx17) [reply .ci/deps.pip.sh]: Ok

File: .ci/deps.sh (hunk 19 above)
Comment (harshhx17) [reply .ci/deps.sh]: This had fixed an issue with the certificates
Will get a link to the error and see if there is a better solution

File: .circleci/config.yml (hunk 20 above)
Comment (harshhx17) [reply .circleci/config.yml]: It was pre-installed in the previous ci image
I will check the details, don't remember them well

File: .circleci/config.yml (hunk 22 above)
Comment (harshhx17) [reply .circleci/config.yml]: I will remove this...just added this because bears having gem dependencies were getting skipped.
which was actually due to the wrong binaries going to the path
Sort of added a redundant line

File: .circleci/config.yml (hunk 23 above)
Comment (harshhx17) [reply .circleci/config.yml]: I wanted to use an image that was having trusty
This is practically the only circleci image that had trusty and had something that was required by us
Python images use stretch, jessie

File: .ci/deps.pip.sh (hunk 17 above)
Comment (harshhx17) [reply .ci/deps.pip.sh]: in circleci 1.0 python 2.7.10 is already installed
And we have a command 
   pyenv local $dep_version 2.7.10
which doesnot work without installing this version
And i thought there might be some portion of code which is dependent on it

Should I just find the default version of python 2 installed and change this command to
   pyenv local $dep_version $installed_version

File: .circleci/config.yml (hunk 20 above)
Comment (harshhx17) [reply .circleci/config.yml]: Will also find the yarn issue and see if this can solve that issue

File: .circleci/config.yml (hunk 21 above)
Comment (harshhx17) [reply .circleci/config.yml]: I remember you had told to remove virtual-env from circleci 2.0 from the coala/coala code base
Wanted to remove this and cache it but some had some problems with it...
Then I thought that I should first get a green build and then try for this

Or can it create a conflict with some packages??
